File: agents/random_agent.cpp
#include "agent.h"

int main() {
    std::srand(std::time(nullptr));

    while (true) {
        GameState gs = read_state();
        std::vector<Action> actions;

        for (auto& piece : gs.my_pieces) {
            // Get all reachable cells for MOVE (radius 2, exclude current)
            auto move_cells = reachable(piece.row, piece.col, 2);
            std::vector<Cell> move_targets;
            for (auto& c : move_cells)
                if (!(c.r == piece.row && c.c == piece.col))
                    move_targets.push_back(c);

            // Get adjacent cells for CLONE (radius 1)
            auto clone_targets = neighbors(piece.row, piece.col);

            bool can_move  = !move_targets.empty();
            bool can_clone = !clone_targets.empty();

            if (!can_move && !can_clone) continue;

            // Random choice: 40% CLONE, 60% MOVE (if both available)
            bool do_clone = can_clone && (!can_move || (std::rand() % 10 < 4));

            if (do_clone) {
                auto& t = clone_targets[std::rand() % clone_targets.size()];
                actions.push_back({piece.id, "CLONE", t.r, t.c});
            } else {
                auto& t = move_targets[std::rand() % move_targets.size()];
                actions.push_back({piece.id, "MOVE", t.r, t.c});
            }
        }

        write_actions(actions);
    }
    return 0;
}

File: agents/random_agent.py
"""Python random agent — for testing the simulator without a C++ compiler."""
import sys
import json
import random
from collections import deque

SIDE = 11

def is_valid(r, c):
    return 0 <= r < SIDE and 0 <= c <= 2 * r

def neighbors(r, c):
    if c % 2 == 0:
        cands = [(r, c-1), (r, c+1), (r+1, c+1)]
    else:
        cands = [(r, c-1), (r, c+1), (r-1, c-1)]
    return [(nr, nc) for nr, nc in cands if is_valid(nr, nc)]

def reachable(r, c, radius):
    visited = {(r, c)}
    q = deque([(r, c, 0)])
    while q:
        cr, cc, d = q.popleft()
        if d >= radius:
            continue
        for nr, nc in neighbors(cr, cc):
            if (nr, nc) not in visited:
                visited.add((nr, nc))
                q.append((nr, nc, d + 1))
    return visited

while True:
    line = sys.stdin.readline()
    if not line:
        break
    state = json.loads(line.strip())
    actions = []
    for piece in state["my_pieces"]:
        r, c = piece["row"], piece["col"]
        move_targets = list(reachable(r, c, 2) - {(r, c)})
        clone_targets = neighbors(r, c)
        if not move_targets and not clone_targets:
            continue
        if clone_targets and random.random() < 0.4:
            t = random.choice(clone_targets)
            actions.append({"piece_id": piece["id"], "action": "CLONE", "row": t[0], "col": t[1]})
        elif move_targets:
            t = random.choice(move_targets)
            actions.append({"piece_id": piece["id"], "action": "MOVE", "row": t[0], "col": t[1]})
    print(json.dumps({"actions": actions}), flush=True)
